### PR TITLE
SavedState Step 7 - Arguments

### PR DESCRIPTION
**Pull request by 0nko**
This is the 7th step of the SavedState module integration.

During the review of the previous PRs, a couple of issues have surfaced relating to the initialization of ViewModels. The fact that the ViewModels must be _started_ and supplied with additional parameters creates a potential breeding ground for bugs :). There are different ways how ViewModels can be (re)started (the first opening of a screen, closing/reopening of the screen, app getting killed & restored) and each scenario must be handled properly so that the correct state is displayed.

The navigation component gives us a nice way to pass type-safe args between fragments. But in reality, these arguments are really meant for the ViewModels. So this PR simplifies things and provides a way to access arguments directly in the ViewModel. A nice bonus is that the arguments are available during the ViewModel creation and they're automatically preserved by the `SavedStateHandle`.

The changes contained in this PR:

- Replaced the use of `SavedStateHandle` with a custom `SavedStateWithArgs` wrapper class
- Elimination of the `start()` function and all `lateinit` properties from the `IssueRefundViewModel`
- The use of `SavedStateWithArgs.navArgs()` delegate for acquiring the arguments in `IssueRefundViewModel`

### CODE REVIEW

**Comment by JavonDavis** [comment]
Hey! Bumping this to `3.2`, if you'd like to get it in for this release please merge this into release/3.1 and then DM me to release a new Beta version.`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductsModule.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.woocommerce.android.di.FragmentScope
+import com.woocommerce.android.ui.products.ProductsModule.ProductDetailFragmentModule
+import com.woocommerce.android.ui.products.ProductsModule.ProductListFragmentModule
+import dagger.Module
+import dagger.android.ContributesAndroidInjector
+
+@Module(includes = [
+    ProductDetailFragmentModule::class,
+    ProductListFragmentModule::class
```
Comment (AmandaRiu) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductsModule.kt]: It looks like the `ProductVariantsFragmentModule::class` is missing...unless I'm not understanding what is happening here 🤷‍♀

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt [hunk 2]
```diff
@@ -79,98 +76,80 @@ class IssueRefundViewModel @AssistedInject constructor(
...
-            }
-        }
+        initRefundByAmountState(order)
+        initRefundSummaryState()
     }
 
-    // Temporary workaround for the activity ViewModel scope
-    private fun resetLiveData() {
+    // temporary workaround for the activity ViewModel scope
+    fun resetLiveData() {
```
Comment (AmandaRiu) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt]: Is this still needed? It doesn't look to be used.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt [hunk 3]
```diff
@@ -237,6 +216,12 @@ class IssueRefundViewModel @AssistedInject constructor(
 
                 // pause here until the snackbar is dismissed to allow for undo action
                 val wasRefundCanceled = waitForCancellation()
+                triggerEvent(ShowSnackbar(
```
Comment (AmandaRiu) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt]: Slick 😍

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductsModule.kt (hunk 1 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductsModule.kt]: You're right, I forgot to include it when I merged the changes from `develop`. Thanks!

**Pull request merged**